merkle/smt/writer.go

```diff
@@ -0,0 +1,162 @@
+		// Check if this ID ends the shard.
+		if next == len(upd) || upd[next].ID.Prefix(top) != pref {
+			shards = append(shards, upd[begin:next])
+			begin = next
+		}
+	}
+	return shards, nil
+}
+
+// Write applies the given list of updates to a single shard, and returns the
+// resulting update of the shard root. It uses the given node accessor for
+// reading and writing tree nodes. Typically, the input updates have been
+// obtained from Split method.
+func (w *Writer) Write(upd []NodeUpdate, acc NodeBatchAccessor) (NodeUpdate, error) {
+	if len(upd) == 0 {
+		return NodeUpdate{}, errors.New("nothing to write")
+	}
+	depth := upd[0].ID.BitLen()
+	top, found := w.depths[depth]
+	if !found {
+		return NodeUpdate{}, fmt.Errorf("unexpected depth %d", depth)
+	}
+
+	hs, err := NewHStar3(upd, w.hasher.HashChildren, depth, top)
+	if err != nil {
+		return NodeUpdate{}, err
+	}
+	ids := hs.Prepare()
+	nodes, err := acc.Get(ids)
+	if err != nil {
+		return NodeUpdate{}, err
+	}
+	sa := w.newAccessor(nodes)
+	topUpd, err := hs.Update(sa)
+	if err != nil {
+		return NodeUpdate{}, err
+	} else if ln := len(topUpd); ln != 1 {
+		return NodeUpdate{}, fmt.Errorf("writing across %d shards, want 1", ln)
+	}
+	if err := acc.Set(sa.writes); err != nil {
+		return NodeUpdate{}, err
+	}
+
+	return topUpd[0], nil
+}
+
+// newAccessor returns a NodeAccessor for HStar3 algorithm based on the set of
+// preloaded node hashes.
+func (w *Writer) newAccessor(nodes map[tree.NodeID2][]byte) *shardAccessor {
+	// For any node that HStar3 reads, it also writes its sibling. Therefore we
+	// can pre-allocate this many items for the writes slice.
+	// TODO(pavelkalinnikov): The actual number of written nodes will be slightly
+	// bigger by at most the number of written leaves. Try allocating precisely.
+	writes := make([]NodeUpdate, 0, len(nodes))
+	return &shardAccessor{w: w, reads: nodes, writes: writes}
+}
+
+// shardAccessor provides read and write access to nodes used by HStar3.
+type shardAccessor struct {
+	w      *Writer
+	reads  map[tree.NodeID2][]byte
+	writes []NodeUpdate
+}
+
+// Get returns the hash of the given node from the preloaded map, or a hash of
+// an empty subtree at this position if such node is not found.
+func (s *shardAccessor) Get(id tree.NodeID2) ([]byte, error) {
+	if hash, ok := s.reads[id]; ok && hash != nil {
+		return hash, nil
+	}
+	return hashEmpty(s.w.hasher, s.w.treeID, id), nil
+}
+
+// Set adds the given node hash update to the list of writes.
+func (s *shardAccessor) Set(id tree.NodeID2, hash []byte) {
+	s.writes = append(s.writes, NodeUpdate{ID: id, Hash: hash})
+}
+
+// TODO(pavelkalinnikov): Make MapHasher.HashEmpty method take the id directly.
+func hashEmpty(hasher hashers.MapHasher, treeID int64, id tree.NodeID2) []byte {
+	index := make([]byte, hasher.Size())
+	copy(index, id.FullBytes())
+	if last, bits := id.LastByte(); bits != 0 {
+		index[len(id.FullBytes())] = last
+	}
+	height := hasher.BitLen() - int(id.BitLen())
```

Comment:
This looks like it's doing what `FullBytes` should be doing? 
Should this be inside of NodeID2?

Reply:
`FullBytes` does not return the last byte, even if it's full, so it's not the same thing.
I don't want to put this to `NodeID2` yet, until it's clear that there is more than one use-case. Also I hope to avoid this conversion entirely when `storage` and `hashers.MapHasher` support `NodeID2`.